### PR TITLE
Issue #383: sftp work

### PR DESCRIPTION
**Pull request by shachiagarwal**
# Please include a summary of the changes

* vane/device_interface.py added transfer_file func for netmiko conn
* vane/tests_tools.py added transfer_file func for TestOps class

# Any specific logic/part of code you need extra attention on

Changes in netmiko conn and TestOps class

# Include the Issue number and link

#383 

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [X] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [ ] Other (please specify)

# Effort required on reviewers end

- - [ ] Easy
- - [X] Medium
- - [ ] Hard 

# How Has This Been Tested?
Added unittests and sample tests. Here is one of the screenshots:
Screenshot 2023-08-17 at 12 48 07 AM

## Bug fix

    Include before and after snapshots/screenshots/changed logs
    
## New feature

    Ensure current test cases pass and include newer test cases if required
    
# CI pipeline result

- - [X] Pass
- - [ ] Fail
  
  If it fails, explain the reason and whether or not we should ignore the failure
  
# Verify Documentation Update

    If applicable, ensure documentation has been updated for ReadMe, Getting Started guide, and Style guide
    
# Additional comments

### CODE REVIEW

File: sample_network_tests/vane_system_tests/test_vane.py [hunk 1]
```diff
@@ -292,3 +293,237 @@ def test_if_setup_fail_is_handled(self):
...
+
+        except (AttributeError, LookupError, EapiError) as exception:
+            logging.error(
+                f"On device {tops.dut_name}: Error while running testcase on DUT is: "
+                f"{str(exception)}"
+            )
+            tops.actual_output = str(exception)
+            tops.output_msg += (
+                f"EXCEPTION encountered on device {tops.dut_name}, while "
+                f"investigating if ssh can be used to run cmds. Vane recorded error: {exception}"
```
Comment (rewati-arista) [inline sample_network_tests/vane_system_tests/test_vane.py:475]: Need an appropriate output message: 
```suggestion
                f"investigating if file transfer using ssh. Vane recorded error: {exception}"
```

File: sample_network_tests/vane_system_tests/test_vane.py [hunk 2]
```diff
@@ -292,3 +293,237 @@ def test_if_setup_fail_is_handled(self):
...
```
Comment (rewati-arista) [inline sample_network_tests/vane_system_tests/test_vane.py:523]: same as above
```suggestion
                f"investigating file transfer using sftp. Vane recorded error: {exception}"
```

File: tests/unittests/test_tests_tools.py [hunk 3]
```diff
@@ -1218,6 +1218,8 @@ def test_test_ops_generate_report(logdebug, mocker):
         "test_steps": [],
         "show_cmds": {"DCBBW1": ["show version", "show version"]},
         "fail_or_skip_reason": "",
+        "cfg_cmd_txts": {"DCBBW1": []},
```
Comment (rewati-arista) [inline tests/unittests/test_tests_tools.py:1221]: now that we got rid of these in https://github.com/aristanetworks/vane/pull/382, do we need line 1221 and 1222?

File: tests/unittests/test_tests_tools.py [hunk 4]
```diff
@@ -1513,6 +1509,247 @@ def test_test_ops_run_show_cmds_text_exception_fail(mocker):
...
+    cfg_cmds = ["interface eth16", "description unittest"]
+
+    actual_output = tops.run_cfg_cmds(cfg_cmds, dut)
+
+    # assert return values
+    assert actual_output == [
+        {},
+        {},
+    ]
+    assert tops._cfg_cmds == {
```
Comment (rewati-arista) [inline tests/unittests/test_tests_tools.py:1557]: these 2 asserts (line 1557, line 1562) need to be changed, should have the changes made in https://github.com/aristanetworks/vane/pull/382

File: tests/unittests/test_tests_tools.py [hunk 5]
```diff
@@ -1513,6 +1509,247 @@ def test_test_ops_run_show_cmds_text_exception_fail(mocker):
...
+    dut["ssh_conn"] = dut["connection"]
+    tops.show_clock_flag = True
+    cfg_cmds = ["interface eth16", "description unittest"]
+
+    actual_output = tops.run_cfg_cmds(cfg_cmds, dut, conn_type="ssh")
+
+    # assert return values
+    assert actual_output == config_return_value
+
+    assert tops._cfg_cmds == {
```
Comment (rewati-arista) [inline tests/unittests/test_tests_tools.py:1609]: these 2 asserts (line 1609, line 1614) need to be changed, should have the changes made in https://github.com/aristanetworks/vane/pull/382

**Review by rewati-arista — CHANGES_REQUESTED** [review]
This PR needs to be merged carefully, and ideally should be merged after merging https://github.com/aristanetworks/vane/pull/382. Since a lot of changed made in 382 have not be reflected here. Have tried flagging those changes in the review. Apart from that rest looks good!

File: sample_network_tests/vane_system_tests/test_vane.py (hunk 1 above)
Comment (shachiagarwal) [reply sample_network_tests/vane_system_tests/test_vane.py:475]: Agreed. I did at one point rebase it #382's  branch. But then I change it again. I will be careful.

**Comment by shachiagarwal** [comment]
Too complex to merge. Created new PR #384

**Pull request closed without merging**